go/types/value_decoder.go

```diff
@@ -89,91 +133,229 @@ func (r *valueDecoder) readValueSequence() ValueSlice {
+	}
+
+	return listLeafSequence{leafSequence{r.vrw, r.byteSlice(start, end), offsets}}
 }
 
-func (r *valueDecoder) readMapLeafSequence() orderedSequence {
-	count := r.readCount()
-	data := []mapEntry{}
-	for i := uint64(0); i < count; i++ {
-		k := r.readValue()
-		v := r.readValue()
-		data = append(data, mapEntry{k, v})
+func (r *valueDecoder) readBlobSequence() sequence {
+	start := r.pos()
+	offsets := []uint32{start}
+	r.skipKind()
+	offsets = append(offsets, r.pos())
+	level := r.readCount()
+	offsets = append(offsets, r.pos())
+	if level > 0 {
+		offsets = append(offsets, r.skipMetaSequence(BlobKind, level)...)
+	} else {
+		offsets = append(offsets, r.skipBlobLeafSequence()...)
 	}
+	end := r.pos()
 
-	return mapLeafSequence{leafSequence{r.vrw, len(data), MapKind}, data}
+	if level > 0 {
+		return metaSequence{r.vrw, r.byteSlice(start, end), offsets}
+	}
+
+	return blobLeafSequence{leafSequence{r.vrw, r.byteSlice(start, end), offsets}}
 }
 
-func (r *valueDecoder) readMetaSequence(k NomsKind, level uint64) metaSequence {
-	count := r.readCount()
+func (r *valueDecoder) readSetSequence() orderedSequence {
+	start := r.pos()
+	offsets := []uint32{start}
+	r.skipKind()
+	offsets = append(offsets, r.pos())
+	level := r.readCount()
+	offsets = append(offsets, r.pos())
+	if level > 0 {
+		offsets = append(offsets, r.skipMetaSequence(SetKind, level)...)
+	} else {
+		offsets = append(offsets, r.skipValueSequence()...)
+	}
+	end := r.pos()
+
+	if level > 0 {
+		return metaSequence{r.vrw, r.byteSlice(start, end), offsets}
+	}
+
+	return setLeafSequence{leafSequence{r.vrw, r.byteSlice(start, end), offsets}}
+}
+
+func (r *valueDecoder) readMapSequence() orderedSequence {
+	start := r.pos()
+	offsets := []uint32{start}
+	r.skipKind()
+	offsets = append(offsets, r.pos())
+	level := r.readCount()
+	offsets = append(offsets, r.pos())
+	if level > 0 {
+		offsets = append(offsets, r.skipMetaSequence(MapKind, level)...)
+	} else {
+		offsets = append(offsets, r.skipMapLeafSequence()...)
+	}
+	end := r.pos()
+
+	if level > 0 {
+		return metaSequence{r.vrw, r.byteSlice(start, end), offsets}
+	}
+
+	return mapLeafSequence{leafSequence{r.vrw, r.byteSlice(start, end), offsets}}
+}
 
-	data := []metaTuple{}
+func (r *valueDecoder) skipList() {
+	r.skipListOrSet(ListKind)
+}
+
+func (r *valueDecoder) skipSet() {
+	r.skipListOrSet(SetKind)
+}
+
+func (r *valueDecoder) skipListOrSet(kind NomsKind) {
+	r.skipKind()
+	level := r.readCount()
+	if level > 0 {
+		r.skipMetaSequence(kind, level)
+	} else {
+		r.skipValueSequence()
+	}
+}
+
+func (r *valueDecoder) skipMap() {
+	r.skipKind()
+	level := r.readCount()
+	if level > 0 {
+		r.skipMetaSequence(MapKind, level)
+	} else {
+		r.skipMapLeafSequence()
+	}
+}
+
+func (r *valueDecoder) skipBlob() {
+	r.skipKind()
+	level := r.readCount()
+	if level > 0 {
+		r.skipMetaSequence(BlobKind, level)
+	} else {
+		r.skipBlobLeafSequence()
+	}
+}
+
+// func (r *valueDecoder) skipListLeafSequence() {
```

Comment:
remove?

Reply:
Done.

Sorry for all the commented out code. I should have gone through this PR a few more times.